src/gui/qgssourceselectprovider.h

```diff
@@ -0,0 +1,71 @@
+/***************************************************************************
+  qgssourceselectprovider.h - QgsSourceSelectProvider
+
+ ---------------------
+ begin                : 1.9.2017
+ copyright            : (C) 2017 by Alessandro Pasotti
+ email                : apasotti at boundlessgeo dot com
+ ***************************************************************************
+ *                                                                         *
+ *   This program is free software; you can redistribute it and/or modify  *
+ *   it under the terms of the GNU General Public License as published by  *
+ *   the Free Software Foundation; either version 2 of the License, or     *
+ *   (at your option) any later version.                                   *
+ *                                                                         *
+ ***************************************************************************/
+#ifndef QGSSOURCESELECTPROVIDER_H
+#define QGSSOURCESELECTPROVIDER_H
+
+
+#include "qgis_gui.h"
+#include "qgis.h"
+#include "qgsguiutils.h"
+#include "qgsproviderregistry.h"
+#include "qgsabstractdatasourcewidget.h"
+
+class QString;
+class QWidget;
+
+/** \ingroup gui
+ * This is the interface for those who want to add entries to the \see QgsDataSourceManagerDialog
+ *
+ * \since QGIS 3.0
+ */
+class GUI_EXPORT QgsSourceSelectProvider
+{
+  public:
+    virtual ~QgsSourceSelectProvider() = default;
+
+    //! Data Provider key
+    virtual QString providerKey() const = 0;
+
+    /** Source select provider name, this is useful to retrieve
+     * a particular source select in case the provider has more
+     * than one, it should be unique among all providers.
+     *
+     * The default implementation returns the providerKey()
+     */
+    virtual QString name() const { return providerKey(); }
+
+    //! Text for the menu item entry, it will be visible to the user so make sure it's translatable
+    virtual QString text() const = 0;
+
+    /** Creates a new instance of an QIcon for the menu item entry
+     * Caller takes responsibility of deleting created.
```

Comment:
This isn't a factory, and doesn't need the 'caller takes responsibility...' docs.